test/e2e/node/gpu.go

```diff
@@ -0,0 +1,206 @@
+		gomega.Expect(log).To(gomega.ContainSubstring("Time taken for 5000x5000 matrix multiplication"))
+	})
+})
+
+func createAndValidatePod(ctx context.Context, f *framework.Framework, podClient *e2epod.PodClient, pod *v1.Pod) {
+	pod = podClient.Create(ctx, pod)
+
+	ginkgo.By("Watching for error events or started pod")
+	ev, err := podClient.WaitForErrorEventOrSuccess(ctx, pod)
+	framework.ExpectNoError(err)
+	gomega.Expect(ev).To(gomega.BeNil())
+
+	ginkgo.By("Waiting for pod completion")
+	err = e2epod.WaitForPodNoLongerRunningInNamespace(ctx, f.ClientSet, pod.Name, f.Namespace.Name)
+	framework.ExpectNoError(err)
+	pod, err = podClient.Get(ctx, pod.Name, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+
+	ginkgo.By("Checking that the pod succeeded")
+	gomega.Expect(pod.Status.Phase).To(gomega.Equal(v1.PodSucceeded))
+}
+
+func testNvidiaCLIPod() *v1.Pod {
+	podName := "gpu-cli-" + string(uuid.NewUUID())
+	pod := v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        podName,
+			Annotations: map[string]string{},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "nvidia-smi",
+					Image: "nvidia/cuda:12.3.2-runtime-ubuntu22.04",
+					Resources: v1.ResourceRequirements{
+						Limits: v1.ResourceList{
+							"nvidia.com/gpu": resource.MustParse("1"),
+						},
+					},
+				},
+			},
+			RestartPolicy: v1.RestartPolicyNever,
+		},
+	}
+	return &pod
+}
+
+func testMatrixMultiplicationPod() *v1.Pod {
+	podName := "gpu-matmul-" + string(uuid.NewUUID())
+	pod := v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        podName,
+			Annotations: map[string]string{},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "gpu-matmul",
+					Image: "tensorflow/tensorflow:latest-gpu",
+					Command: []string{
+						"python",
+						"-c",
+						`
+import tensorflow as tf
+import time
+
+print("TensorFlow version:", tf.__version__)
+print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))
+
+# Simple matrix multiplication test
+with tf.device('/GPU:0'):
+    a = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+    b = tf.constant([[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]])
+    c = tf.matmul(a, b)
+
+print("Matrix multiplication result:", c.numpy())
+
+# Performance test
+n = 5000
+start_time = time.time()
+with tf.device('/GPU:0'):
+    matrix1 = tf.random.normal((n, n))
+    matrix2 = tf.random.normal((n, n))
+    result = tf.matmul(matrix1, matrix2)
+end_time = time.time()
+
+print(f"Time taken for {n}x{n} matrix multiplication: {end_time - start_time:.2f} seconds")
+`,
+					},
+					Resources: v1.ResourceRequirements{
+						Limits: v1.ResourceList{
+							"nvidia.com/gpu": resource.MustParse("1"),
+						},
+					},
+				},
+			},
+			RestartPolicy: v1.RestartPolicyNever,
+		},
+	}
+	return &pod
+}
+
+func checkEnvironmentAndSkipIfNeeded(ctx context.Context, clientSet clientset.Interface) {
```

Comment:
Sorry, I cannot find where is the driver being installed. Or we assume the environment has a driver installed?

Reply:
/hold

to understand what we test

Reply:
@SergeyKanzhelev `kubetest2 ec2` has an option
https://cs.k8s.io/?q=DevicePluginNvidia%7CENABLE_NVIDIA_DEVICE_PLUGIN&i=nope&files=&excludeFiles=&repos=

Reply:
xref: https://github.com/kubernetes/test-infra/pull/33492

Reply:
I wonder if we need to rely on device driver that comes with the platform or should have it installed by test. If installed by test - we will have more explicit statement on what nvidia drivers k8s is compatible with.

Reply:
@SergeyKanzhelev i am happy whichever way you want to go with. The main focus of this PR to get something up which we can then tweak as we are blind right now. Once i get this working on AWS, we can figure out whether we want to do the GCP one via a change in kubeup or e2e-test code and then i can fix up aws side to match.

Reply:
ok, makes sense

/unhold

Reply:
thanks @SergeyKanzhelev